python/tests/ibd.py

```diff
@@ -0,0 +1,421 @@
+        self.current_parent = self.ts.tables.edges.parent[0]
+        self.current_time = 0
+        self.A = [None for _ in range(ts.num_nodes)]  # Descendant segments
+        self.tables = self.ts.tables
+        self.parent_list = [[] for i in range(0, self.ts.num_nodes)]
+        for e in self.ts.tables.edges:
+            if (
+                len(self.parent_list[e.child]) == 0
+                or e.parent != self.parent_list[e.child][-1]
+            ):
+                self.parent_list[e.child].append(e.parent)
+        # Objects below are needed for the IBD segment-holding object.
+        self.sample_id_map = self.get_sample_id_map()
+        self.num_samples = len(self.samples)
+        self.sample_pairs = self.get_sample_pairs()
+        self.PARENT_SHOULD_BE_ADDED = True
+
+        # Note: in the C code the object below should be a struct array.
+        # Each item will be accessed using its index, which corresponds to a particular
+        # sample pair. The mapping between index and sample pair is defined in the
+        # find_sample_pair_index method further down.
+
+        self.ibd_segments = {}
+        for key in self.sample_pairs:
+            self.ibd_segments[key] = None
+
+    def add_ibd_segments(self, sample0, sample1, seg):
+        index = self.find_sample_pair_index(sample0, sample1)
+
+        # Note: the code below is specific to the Python implementation, where the
+        # output is a dictionary indexed by sample pairs.
+        # In the C implementation, it'll be more like
+        # self.ibd_segments[index].add(seg)
+
+        if self.ibd_segments[self.sample_pairs[index]] is None:
+            self.ibd_segments[self.sample_pairs[index]] = SegmentList(
+                head=seg, tail=seg
+            )
+        else:
+            self.ibd_segments[self.sample_pairs[index]].add(seg)
+
+    def get_sample_id_map(self):
+        """
+        Returns id_map, a vector of length ts.num_nodes. For a node with id i,
+        id_map[i] is the position of the node in self.samples. If i is not a
+        user-specified sample, id_map[i] is -1.
+        Note: it assumes nodes in the TS are numbered 0 to ts.num_nodes - 1
+        """
+        id_map = [-1] * self.ts.num_nodes
+        for i, samp in enumerate(self.samples):
+            id_map[samp] = i
+
+        return id_map
+
+    def get_sample_pairs(self):
+        """
+        Returns a list of all pairs of samples. Replaces itertools.combinations.
+        Note: they must be sorted
+        """
+        sample_pairs = []
+        for ind, i in enumerate(self.samples):
+            for j in self.samples[(ind + 1) :]:
+                sample_pairs.append((i, j))
+
+        return sample_pairs
+
+    def find_sample_pair_index(self, sample0, sample1):
+        """
+        Note: this method isn't strictly necessary for the Python implementation
+        but is needed for the C implemention, where the output ibd_segments is a
+        struct array.
```

Comment:
I'm still worried about the N^2 here. It might be easier all around, and avoid N^2, to have the output be a dictionary indexed by a sorted tuple of the sample IDs; and in C we could use an avl_tree, which does more or less the same thing.

Reply:
We can't use an AVL tree here unfortunately @petrelharp  - the one we're using in msprime is GPL. So, I was thinking that the linear map would be simpler for now. I agree that given the list is likely to be sparse this N^2 implementation will be wasteful of memory. I doubt the time required to go through the output to find the non-null entries will end up being significant.

I agree that ultimately we will need to have some sort of sparse representation for the output, which will be keyed by the sorted pair (or equivalently, the index computed here). We will have to make some technology choices about which AVL tree/hash table implementation to use though, which I don't want to get into here.

Reply:
Ok - that'll be easy enough to change, but I think we'll have to do it before applying this to the big datasets.